Moblin/HaishinKit/Media/AudioUnit.swift

```diff
@@ -19,26 +19,120 @@ func makeChannelMap(
     return result.map { NSNumber(value: $0) }
 }
 
+protocol ReplaceAudioSampleBufferDelegate: AnyObject {
+    func didOutputReplaceSampleBuffer(cameraId: UUID, sampleBuffer: CMSampleBuffer)
+}
+
 private class ReplaceAudio {
-    var nextPresentationTimeStamp: CMTime = .zero
+    private var cameraId: UUID
+    private var latency: Double
+    private var frameRate: Double = 0.0
+    private var sampleBufferQueue: [CMSampleBuffer] = []
+    private var startTime: Double?
+    private var state: State = .initializing
+    private var initializationDuration: Double = 1
+    private var outputTimer: DispatchSourceTimer?
+    private enum State {
+        case initializing
+        case buffering
+        case outputting
+    }
+
+    private var maxQueueSize: Int {
+        return Int(latency * frameRate + 1)
+    }
+
+    private var minQueueSize: Int {
+        return Int(latency * frameRate - 1)
+    }
+
+    weak var delegate: ReplaceAudioSampleBufferDelegate?
+
+    init(cameraId: UUID, latency: Double) {
+        self.cameraId = cameraId
+        self.latency = 1 + latency
+    }
+
+    func appendSampleBuffer(_ sampleBuffer: CMSampleBuffer) {
+        let currentTime = CACurrentMediaTime()
+        if startTime == nil {
+            startTime = currentTime
+        }
+        sampleBufferQueue.append(sampleBuffer)
+        // logger.info("ReplaceAudio Queue Count: \(sampleBufferQueue.count)")
+
+        switch state {
+        case .initializing:
+            // logger.info("ReplaceVideo initializing.")
+            if currentTime - startTime! >= initializationDuration {
+                initialize(sampleBuffer: sampleBuffer)
```

Comment:
What is TCP has issues and only two frames have been received after 1 second? A bit too fragile I think.

Reply:
So you would make it more dependent on the count of frames?

Reply:
Sounds better, yes. The alternative is to set the timer interval to match the stream output fps.